### PR TITLE
Port the core/code block to be used by the ReactNative mobile GB app

### PR DESCRIPTION
**Pull request by hypest**
## Description
This PR introduces the minimal set of changes to make the core/code block ready to be consumed by a ReactNative app, in the road to implement a "native" mobile GB editing experience as per [the Make post](https://make.wordpress.org/mobile/2018/03/21/gutenberg-on-mobile/).

Among the goals is to eventually be able to introduce patterns for how to share code between GB the webapp and GB the native mobile app. Please, feel free to open PRs against this one with proposals towards that goal.

## TODO before lifting the "In progress" label:
- [x] Make sure the (web) GB build and webapp work normally
- [x] Factorize the RN platform dependent files to minimize code duplication between Android and iOS
- [x] ~Have a proposal ready for how to change the `core/code` block definition so its high level definition is shared among the 3 platforms (web, Android, iOS)~
For the time being, the `.native.js` variants of key files (`index.js` among which is the most important ones) will serve as the method we provide compatibility with an RN app. The goal is still to come up with a "common language" down the road but we need to port more code before patterns on what the common language is emerge.

## Types of changes
* What types of changes does your code introduce?

Introduces `.native.js` variants of various key files. Those are picked up instead of the corresponding `.js` files when the code is imported in an RN build. The regular Gutenberg build just ignores the `.native.js` files.

The specialized `index.native.js` files play a key role: They are importing only the submodules needed by the RN implementation. These submodules will gradually be ported and only where appropriate.

* New feature (non-breaking change which adds functionality)

Ports the `core/code` block into a RN compatible component.

* Breaking change (fix or feature that would cause existing functionality to not work as expected)

None.

## Checklist:
- [x] My code is tested. (Yes, [tests are on the gutenberg-mobile repo](https://github.com/wordpress-mobile/gutenberg-mobile/pull/25))
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by hypest** [comment]
👋 @gziolo , please feel free to add more reviewers to this PR if you think it's good they keep an open eye to this PR. Thanks!

👋@maxme , feel free to add more `TODO` items to the list. Thanks!

**Comment by aduth** [comment]
>Have a proposal ready for how to change the core/code block definition so it's high level definition is shared among the 3 platforms (web, Android, iOS)

May also tie into #2751 (registration of said high-level definition on the server).

**Comment by gziolo** [comment]
> >Have a proposal ready for how to change the core/code block definition so it's high level definition is shared among the 3 platforms (web, Android, iOS)
> 
> May also tie into #2751 (registration of said high-level definition on the server).

In general, we are looking into having a way to define all blocks on the server (minus JS specific bits like `edit`, `save` or other function based properties). We will have to wrap this with REST API endpoint to make it easy to consume by the platforms that don't use HTML to expose data from the server.

**Comment by hypest** [comment]
Collapsed (with 8ee1a4e) platform specific code into `.native.js` files instead of individual (android, ios) ones. Thanks @gziolo for the tip!

**Comment by hypest** [comment]
Update:

* Merged from master and fixed any conflict
* Lifted the #pragma override of the JSX Babel transformer plugin
* Added the mobile native source files (`.native.js` and friends) to the eslint ignore list. Those files are currently processed via Flow and Prettier on the RN project side.
* Tests now pass

**Comment by hypest** [comment]
Update:

* Re-established eslint for `.native.js` files and fixed all error occurrences
* Removed the React import from the `.native.js` files and instead, auto-importing React via the RN project's build pipeline

**Comment by hypest** [comment]
Update: added a commit (685d178) that factors out the _view_ related definition out of the core/code block, into platform specific modules.

Can't tell yet if that's going to be scalable or usable in the rest of the blocks but, would love to hear some initial comments/thoughts on this.

WDYT @gziolo ?

**Comment by maxme** [comment]
Note: I changed the base branch to point to our future feature branch: `rnmobile/react-native-mobile-integration-first-steps`

**Comment by hypest** [comment]
After some discussion with @gziolo I've :
* cleaned up the "todo-looking" comments
* minimized the PR even further
* changed the target branch back to `master`
* updated from `master` and fixed any incompatibilities (e.g. due to file structure change)
* updated the PR description
* lifted the `[Status] In Progress` label.

File: core-blocks/code/edit.native.js
Comment (gziolo) [inline core-blocks/code/edit.native.js]: It can be exported as `export default function ...`.

File: core-blocks/code/index.js
Comment (gziolo) [inline core-blocks/code/index.js]: This comment can be removed as we should probably do the same for all `edit` implementations :)

`edit` can be exported as default to have:

```js
import edit from './edit';

// and later
{ 
     ...
    edit,
    ...
}
```

File: core-blocks/code/edit.native.js
Comment (gziolo) [inline core-blocks/code/edit.native.js]: Shouldn't be `style` applied to `View` rather than `PlainText` to mirror web implementation?

File: blocks/plain-text/style.native.scss
Comment (gziolo) [inline blocks/plain-text/style.native.scss]: Did you check if it isn't included in web build by accident?

**Review by gziolo — COMMENTED** [review]
I left some minor comments. In general, I'm okey with the proposed solution, because everything happens in RN specific files. I think it would be a good idea to move all `edit` components to their own files anyway. We can help with that if others agree that it is a way to move forward.

@mtias, @aduth, @youknowriad any thoughts?

File: core-blocks/code/edit.native.js
Comment (hypest) [reply core-blocks/code/edit.native.js]: Makes sense, done with 0edfb18.

File: core-blocks/code/index.js
Comment (hypest) [reply core-blocks/code/index.js]: The comment can be considered trivial if one reads the PR and has an understanding of the `.native.js` pattern. I had added it to help people pick up the pattern but, i'm perfectly fine with removing it 👍 . Done with 9b68ba3.

File: core-blocks/code/edit.native.js
Comment (hypest) [reply core-blocks/code/edit.native.js]: Yes and no. Unfortunately, styling in RN is not cascading by default and we will need more work on establishing a pattern on how to pass and style the blocks/components/containers.

In this particular case for example, the style includes the `fontFamily` property which is supported by the `PlaintText` component (because it's a `TextInput` under the hood) but by the `View` component.

All in all, styling should be considered `alpha` at this stage as well and i'd suggest we work on it in a different PR, if that's OK with you too @gziolo . I'll update the PR description to include some relative info.

File: blocks/plain-text/style.native.scss
Comment (hypest) [reply blocks/plain-text/style.native.scss]: Not extensively, I didn't. What I only tested was modifying values in the style file and see if GB-web picks those up, visually. Tried out adding a `background-color` for example. Any practical way to test it out more thoroughly perhaps?

**Comment by hypest** [comment]
Thanks for the review @gziolo , ready for another pass! FYI, I made a minor change in the meantime (5c93331).

File: core-blocks/code/edit.native.js
Comment (gziolo) [reply core-blocks/code/edit.native.js]: That would be awesome to have it documented in the code next to RN version of the `edit` component.

File: blocks/plain-text/style.native.scss
Comment (gziolo) [reply blocks/plain-text/style.native.scss]: I can test it myself before merging, but I think if you put `background-color: pink` in the RN version, it shouldn't get applied in the web version. That's all we need to ensure :)

Comment (hypest) [reply blocks/plain-text/style.native.scss]: That's grand, that was the test I did and it succeeded (color wasn't picked up) so, feel free to test is your side as well, thanks!

File: core-blocks/code/edit.native.js
Comment (hypest) [reply core-blocks/code/edit.native.js]: Done with 0377c73.

**Review by gziolo — APPROVED** [review]
I tested web version and it still works properly. I would defer the final decision to @aduth and @youknowriad, but speaking myself it's good to go.

**Comment by gziolo** [comment]
Let's get it and iterate with further PRs.

**Comment by mtias** [comment]
Good to see this happening :)

**Pull request merged**
